### PR TITLE
seed/seedwriter,snap/naming: support classic models 

### PR DESCRIPTION
**Pull request by pedronis**
This also adds support for implicit core/snapd for those (and also other cases for core).

Introduces SnapSet.Empty as well.

### CODE REVIEW

**Comment by pedronis** [comment]
based on #7468

File: seed/seedwriter/writer.go [hunk 1]
```diff
@@ -557,66 +599,96 @@ func (w *Writer) resolveChannel(whichSnap string, modSnap *asserts.ModelSnap, op
...
+// Downloaded checks the downloaded snaps metadata provided via
+// setting it into the SeedSnaps returned by the previous
+// SnapsToDownload.  It also returns whether the seed snap set is
+// complete or SnapsToDownload should be called again.
+func (w *Writer) Downloaded() (complete bool, err error) {
+	if err := w.checkStep(downloadedStep); err != nil {
+		return false, err
+	}
+
+	// TODO: w.policy.resetChecks()
```
Comment (stolowski) [inline seed/seedwriter/writer.go]: Same question re a follow up?

File: seed/seedwriter/seed16.go [hunk 2]
```diff
@@ -52,25 +53,27 @@ func (pol *policy16) checkSnapChannel(_ channel.Channel, whichSnap string) error
 	return nil
 }
 
+func makeSystemSnap(snapName string) *asserts.ModelSnap {
+	// TODO: set SnapID too
```
[the remark this answers is not in the document]
Comment (pedronis) [reply seed/seedwriter/seed16.go]: yes, but in these PRs only XXX means something that needs to and I will  fix soon. These instead are real TODOs it the sense that they are not about old features but new things/improvements.

File: seed/seedwriter/seed16.go [hunk 3]
```diff
@@ -108,6 +111,35 @@ func (pol *policy16) checkBase(info *snap.Info, availableSnaps *naming.SnapSet)
...
 
+func (pol *policy16) needsImplicitSnaps(availableSnaps *naming.SnapSet) (bool, error) {
+	// do we need to add implicitly either snapd (or core)
+	hasCore := availableSnaps.Contains(naming.Snap("core"))
+	if len(pol.needsCore) != 0 && !hasCore {
+		// XXX warning on Core 18
+		return true, nil
+	}
+
+	if len(pol.needsCore16) != 0 && !hasCore {
```
Comment (bboozzoo) [inline seed/seedwriter/seed16.go]: Should this check for core16 too?
```
hasCore16 := availableSnaps.Contains(naming.Snap("core16"))
if len(pol.needsCore16) != 0 && (!hasCore || !hasCore16) { ..
```
Although not sure whether seed with core16 would actually be useful.

File: seed/seedwriter/seed16.go [hunk 4]
```diff
@@ -108,6 +111,35 @@ func (pol *policy16) checkBase(info *snap.Info, availableSnaps *naming.SnapSet)
...
+
+	if pol.model.Classic() && !availableSnaps.Empty() {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (pol *policy16) implicitSnaps(availableSnaps *naming.SnapSet) []*asserts.ModelSnap {
+	if len(pol.needsCore) != 0 && !availableSnaps.Contains(naming.Snap("core")) {
```
Comment (bboozzoo) [inline seed/seedwriter/seed16.go]: Do we need to look at pol.needsCore16 too?

File: seed/seedwriter/seed16.go (hunk 3 above)
Comment (pedronis) [reply seed/seedwriter/seed16.go]: so far core16 never made it past beta, and at this point is a unclear when/if that happens.  The *implicit* code here corresponds to this code in image.go:

https://github.com/snapcore/snapd/blob/master/image/image.go#L485
https://github.com/snapcore/snapd/blob/master/image/image.go#L503

File: seed/seedwriter/seed16.go (hunk 4 above)
Comment (pedronis) [reply seed/seedwriter/seed16.go]: this is https://github.com/snapcore/snapd/blob/master/image/image.go#L485

File: seed/seedwriter/seed16.go (hunk 3 above)
Comment (bboozzoo) [reply seed/seedwriter/seed16.go]: Makes sense, we can always revisit if there's ever a need to do so.

**Pull request merged**